hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/table/upgrade/TestUpgradeDowngradeFixtures.java

```diff
@@ -0,0 +1,633 @@
+        tableConfig.getString(HoodieTableConfig.TIMELINE_PATH),
+        "Timeline path should have default value");
+    
+    // Record merge mode should be set
+    assertTrue(tableConfig.contains(HoodieTableConfig.RECORD_MERGE_MODE),
+        "Record merge mode should be set for V8");
+    RecordMergeMode mergeMode = tableConfig.getRecordMergeMode();
+    assertNotNull(mergeMode, "Merge mode should not be null");
+    
+    // Record merge strategy ID should be set
+    assertTrue(tableConfig.contains(HoodieTableConfig.RECORD_MERGE_STRATEGY_ID),
+        "Record merge strategy ID should be set for V8");
+    
+    // Payload class should be set
+    assertTrue(tableConfig.contains(HoodieTableConfig.PAYLOAD_CLASS_NAME),
+        "Payload class should be set for V8");
+    
+    // Initial version should be set
+    assertTrue(tableConfig.contains(HoodieTableConfig.INITIAL_VERSION),
+        "Initial version should be set for V8");
+    
+    // Key generator type should be set if key generator class is present
+    if (tableConfig.contains(HoodieTableConfig.KEY_GENERATOR_CLASS_NAME)) {
+      assertTrue(tableConfig.contains(HoodieTableConfig.KEY_GENERATOR_TYPE),
+          "Key generator type should be set when key generator class is present");
+    }
+  }
+
+  /**
+   * Validate properties for version 9 (index version population).
+   */
+  private void validateVersion9Properties(HoodieTableMetaClient metaClient, HoodieTableConfig tableConfig) {
+    validateVersion8Properties(tableConfig);
+    
+    // Check if index metadata exists and has proper version information
+    Option<HoodieIndexMetadata> indexMetadata = metaClient.getIndexMetadata();
+    if (indexMetadata.isPresent()) {
+      indexMetadata.get().getIndexDefinitions().forEach((indexName, indexDef) -> {
+        assertNotNull(indexDef.getVersion(), 
+            "Index " + indexName + " should have version information in V9");
+      });
+    }
+  }
+
+  /**
+   * Validate properties added during upgrade operations.
+   */
+  private void validateUpgradeProperties(HoodieTableVersion fromVersion, HoodieTableVersion toVersion,
+                                       HoodieTableMetaClient metaClient, HoodieTableConfig tableConfig) {
+    LOG.debug("Validating upgrade properties: {} -> {}", fromVersion, toVersion);
+    
+    // Validate timeline layout changes during upgrades
+    if (fromVersion.versionCode() < HoodieTableVersion.EIGHT.versionCode() 
+        && toVersion.versionCode() >= HoodieTableVersion.EIGHT.versionCode()) {
+      // Timeline layout should change from V1 to V2 during upgrade to V8+
+      Option<TimelineLayoutVersion> layoutVersion = tableConfig.getTimelineLayoutVersion();
+      assertTrue(layoutVersion.isPresent(), "Timeline layout version should be present after upgrade to V8+");
+      assertEquals(TimelineLayoutVersion.LAYOUT_VERSION_2, layoutVersion.get(),
+          "Timeline layout should be upgraded to V2 during upgrade to V8+");
```

Comment:
Move this to `validateVersion8Properties `?

Reply:
Will do so